### PR TITLE
Move Projection Node to SPI

### PR DESCRIPTION
**Pull request by sachdevs**
In preparation for aggregation/projection pushdown, move ProjectNode plan node to SPI. As ProjectNode uses the Assignments class, I had to consider moving it to SPI as well. To me this move makes sense, however let me know if I should be doing this in an alternate way.

AggregationNode will be added to SPI in a separate PR as it is slightly more complicated to refactor it.

TODO: ~~depending on what we do with the Assignments class, move the Assignments test into the SPI test directory.~~ Confirmed with James, we do not want to do this.

```
== RELEASE NOTES ==
General 
* Move `ProjectNode` to SPI. Connectors can now pushdown project to table scan.
```

### CODE REVIEW

**Comment by highker** [comment]
Make sure to take a look at https://github.com/prestodb/presto/pull/13045 and go through the comments to make sure the dependency and logic is clean.

**Comment by sachdevs** [comment]
Thanks this will be helpful.

File: presto-spi/src/main/java/com/facebook/presto/spi/plan/ProjectNode.java [hunk 1]
```diff
@@ -74,14 +71,23 @@ public PlanNode getSource()
...
     public PlanNode replaceChildren(List<PlanNode> newChildren)
     {
-        return new ProjectNode(getId(), Iterables.getOnlyElement(newChildren), assignments);
+        requireNonNull(newChildren, "newChildren list is null");
+        checkArgument(newChildren.size() == 1, "newChildren list has multiple items");
+        return new ProjectNode(getId(), newChildren.get(0), assignments);
+    }
+
+    private static void checkArgument(boolean test, String errorMessage)
+    {
```
Comment (sachdevs) [inline presto-spi/src/main/java/com/facebook/presto/spi/plan/ProjectNode.java]: This code is duplicated all over the place in SPI. Right now I copied convention and duplicated it again however, there should be a single static class or something that takes care of argument/state validation.

Here is a list of existing files with this code in SPI (they might be slightly different in behavior from one another but it's near identical)

1. com.facebook.presto.spi.plan.TopNNode
2. com.facebook.presto.spi.plan.LimitNode
3. com.facebook.presto.spi.plan.TableScanNode

And now:

1. ProjectNode.java
2. Assignments.java

Let me know if this is a worthwhile refactor or not.

File: presto-spi/src/main/java/com/facebook/presto/spi/plan/ProjectNode.java [hunk 2]
```diff
@@ -11,25 +11,22 @@
...
 import static java.util.Objects.requireNonNull;
 
 @Immutable
-public class ProjectNode
-        extends InternalPlanNode
+public final class ProjectNode
+        extends PlanNode
 {
     private final PlanNode source;
     private final Assignments assignments;
```
Comment (sachdevs) [inline presto-spi/src/main/java/com/facebook/presto/spi/plan/ProjectNode.java]: See the TODO below this comment on line 34 (idk why github wont let me comment on lines that are not part of the diff).

Should that TODO be addressed before we move ProjectNode into SPI?

File: presto-spi/src/main/java/com/facebook/presto/spi/plan/ProjectNode.java (hunk 1 above)
Comment (highker) [reply presto-spi/src/main/java/com/facebook/presto/spi/plan/ProjectNode.java]: Check comment https://github.com/prestodb/presto/pull/13045#issuecomment-508021391

Comment (hellium01) [reply presto-spi/src/main/java/com/facebook/presto/spi/plan/ProjectNode.java]: Yes, please use duplicated private methods.

File: presto-spi/src/main/java/com/facebook/presto/spi/plan/ProjectNode.java (hunk 2 above)
Comment (hellium01) [reply presto-spi/src/main/java/com/facebook/presto/spi/plan/ProjectNode.java]: Right now, it might be a little hard. Ideally, we should not put identity assignment in map if an output variable is assigned to an identical variable (i.e. `a = a`). However, we currently have two representations of variable now before the whole migration completes: `SymbolReference` and `VariableReference`, which will makes deciding which form to return very difficult.

Personally I don't feel strongly we should inline the `Assignment` into `ProjectNode` like `TableScanNode`. It is slightly cleaner but you will need to refactor a lot of code depends on it.

File: presto-spi/src/main/java/com/facebook/presto/spi/plan/ProjectNode.java (hunk 1 above)
Comment (sachdevs) [reply presto-spi/src/main/java/com/facebook/presto/spi/plan/ProjectNode.java]: Yeah I understand that. However, I feel like since then, the code has been duplicated in more places. Does it make sense to put it in PlanNode itself then? Since the functionality is duplicated in several subclasses.

File: presto-spi/src/main/java/com/facebook/presto/spi/plan/Assignments.java [hunk 3]
```diff
@@ -72,12 +74,12 @@ public static Assignments of(VariableReferenceExpression variable1, RowExpressio
...
     public Assignments(@JsonProperty("assignments") Map<VariableReferenceExpression, RowExpression> assignments)
     {
-        this.assignments = ImmutableMap.copyOf(requireNonNull(assignments, "assignments is null"));
+        this.assignments = unmodifiableMap(new HashMap<>(requireNonNull(assignments, "assignments is null")));
     }
 
     public List<VariableReferenceExpression> getOutputs()
     {
-        return ImmutableList.copyOf(assignments.keySet());
+        return unmodifiableList(new ArrayList<>(assignments.keySet()));
```
Comment (hellium01) [inline presto-spi/src/main/java/com/facebook/presto/spi/plan/Assignments.java]: I am still a little bit concerned about `getOutput` here can have random order. 
It might be ok during optimization, since we will always lookup by variable name. But during execution, all the variables will be changed to `inputChannel` (i.e. the index in the list returned by getOutputs), so a change in order will cause mismatching. `HashMap`'s `keySet` order can easily change. 

Also, the old code 
```
ImmutableList.copyOf(assignments.keySet());
```
is cheap because `ImmutableList.copyOf` smartly utilized underlying data structure, so no copy actually happened. But, we will do copying now each time. 

Here, probably let's save it to member variable during construction and return its value instead. We needed extra memory but it is not as bad as unexpected behavior and GC.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/SetOperationNode.java [hunk 4]
```diff
@@ -111,6 +111,7 @@ protected SetOperationNode(
 
         return builder.build();
     }
+
```
Comment (hellium01) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/SetOperationNode.java]: Unrelated change?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/sanity/TestVerifyNoOriginalExpression.java [hunk 5]
```diff
@@ -74,6 +74,7 @@
     private PlanBuilder builder;
     private ValuesNode valuesNode;
     private CallExpression comparisonCallExpression;
+
```
Comment (hellium01) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/sanity/TestVerifyNoOriginalExpression.java]: Unrelated change?

File: presto-spi/src/main/java/com/facebook/presto/spi/plan/ProjectNode.java [hunk 6]
```diff
@@ -65,7 +61,7 @@ public Assignments getAssignments()
     @Override
     public List<PlanNode> getSources()
     {
-        return ImmutableList.of(source);
+        return Collections.singletonList(source);
```
Comment (hellium01) [inline presto-spi/src/main/java/com/facebook/presto/spi/plan/ProjectNode.java]: static import

File: presto-spi/src/main/java/com/facebook/presto/spi/plan/Assignments.java (hunk 3 above)
Comment (sachdevs) [reply presto-spi/src/main/java/com/facebook/presto/spi/plan/Assignments.java]: Understood, nice catch!

**Comment by sachdevs** [comment]
Fixing any CI tests that may have failed now.

File: presto-spi/src/main/java/com/facebook/presto/spi/plan/Assignments.java [hunk 7]
```diff
@@ -68,16 +70,18 @@ public static Assignments of(VariableReferenceExpression variable1, RowExpressio
...
 
     private final Map<VariableReferenceExpression, RowExpression> assignments;
+    private final ArrayList<VariableReferenceExpression> outputs;
 
     @JsonCreator
     public Assignments(@JsonProperty("assignments") Map<VariableReferenceExpression, RowExpression> assignments)
     {
-        this.assignments = ImmutableMap.copyOf(requireNonNull(assignments, "assignments is null"));
+        this.assignments = unmodifiableMap(new HashMap<>(requireNonNull(assignments, "assignments is null")));
+        this.outputs = new ArrayList<>(assignments.keySet());
```
Comment (hellium01) [inline presto-spi/src/main/java/com/facebook/presto/spi/plan/Assignments.java]: nit: why not put `unmodifiableList` here.

**Comment by hellium01** [comment]
`TestAssingments` probably needs to be updated with the change.

**Comment by sachdevs** [comment]
Fixed CI tests, waiting on pass. Other than that lmk if any other comments.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/plan/TestAssignments.java [hunk 8]
```diff
@@ -13,23 +13,27 @@
...
-public class TestAssingments
+public class TestAssignments
 {
     private final Assignments assignments = assignment(new VariableReferenceExpression("test", BIGINT), TRUE_LITERAL);
 
-    @Test
+    @Test(expectedExceptions = {UnsupportedOperationException.class})
     public void testOutputsImmutable()
     {
-        assertTrue(assignments.getOutputs() instanceof ImmutableCollection);
```
Comment (highker) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/plan/TestAssignments.java]: Rather than throwing an exception, use `UnmodifiableCollection`

File: presto-spi/src/main/java/com/facebook/presto/spi/plan/Assignments.java [hunk 9]
```diff
@@ -209,4 +212,11 @@ public Assignments build()
             return new Assignments(assignments);
         }
     }
+
+    private static void checkState(boolean test, String toFormat, Object... arguments)
```
Comment (highker) [inline presto-spi/src/main/java/com/facebook/presto/spi/plan/Assignments.java]: Inline this helper given it's only used once

File: presto-main/src/test/java/com/facebook/presto/sql/planner/plan/TestAssignments.java (hunk 8 above)
Comment (sachdevs) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/plan/TestAssignments.java]: So there's no "easy" way to check if a collection is an UnmodifiableCollection since the function unmodifiableList is public however the static class UnmodifiableList is package private. So we can use java reflection to test this (as opposed to instanceof):

```
assertTrue(assignments.getOutputs().getClass().isInstance(unmodifiableList(new ArrayList<>())));
```

Which seemed a little ugly so I chose to throw an exception.

Here is an example in SPI that does a similar thing: https://github.com/prestodb/presto/blob/master/presto-spi/src/test/java/com/facebook/presto/spi/predicate/TestEquatableValueSet.java#L299

Comment (highker) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/plan/TestAssignments.java]: Good point. Keep as is then.

**Comment by sachdevs** [comment]
Previous CI failure seems to be a flake, rebasing.
